perception/detected_object_validation/src/obstacle_pointcloud_based_validator.cpp

```diff
@@ -0,0 +1,313 @@
+  const pcl::PointCloud<pcl::PointXY>::Ptr & polygon)
+{
+  if (object.shape.type == Shape::BOUNDING_BOX) {
+    const auto & pose = object.kinematics.pose_with_covariance.pose;
+    double yaw = tf2::getYaw(pose.orientation);
+    Eigen::Matrix2d rotation;
+    rotation << std::cos(yaw), -std::sin(yaw), std::sin(yaw), std::cos(yaw);
+    Eigen::Vector2d offset0, offset1, offset2, offset3;
+    offset0 = rotation *
+              Eigen::Vector2d(object.shape.dimensions.x * 0.5f, object.shape.dimensions.y * 0.5f);
+    offset1 = rotation *
+              Eigen::Vector2d(object.shape.dimensions.x * 0.5f, -object.shape.dimensions.y * 0.5f);
+    offset2 = rotation *
+              Eigen::Vector2d(-object.shape.dimensions.x * 0.5f, -object.shape.dimensions.y * 0.5f);
+    offset3 = rotation *
+              Eigen::Vector2d(-object.shape.dimensions.x * 0.5f, object.shape.dimensions.y * 0.5f);
+    polygon->push_back(
+      pcl::PointXY(toPCL(pose.position.x + offset0.x(), pose.position.y + offset0.y())));
+    polygon->push_back(
+      pcl::PointXY(toPCL(pose.position.x + offset1.x(), pose.position.y + offset1.y())));
+    polygon->push_back(
+      pcl::PointXY(toPCL(pose.position.x + offset2.x(), pose.position.y + offset2.y())));
+    polygon->push_back(
+      pcl::PointXY(toPCL(pose.position.x + offset3.x(), pose.position.y + offset3.y())));
+  } else if (object.shape.type == Shape::CYLINDER) {
+    const auto & center = object.kinematics.pose_with_covariance.pose.position;
+    const auto & radius = object.shape.dimensions.x * 0.5;
+    constexpr int n = 6;
+    for (int i = 0; i < n; ++i) {
+      Eigen::Vector2d point;
+      point.x() = std::cos(
+                    (static_cast<double>(i) / static_cast<double>(n)) * 2.0 * M_PI +
+                    M_PI / static_cast<double>(n)) *
+                    radius +
+                  center.x;
+      point.y() = std::sin(
+                    (static_cast<double>(i) / static_cast<double>(n)) * 2.0 * M_PI +
+                    M_PI / static_cast<double>(n)) *
+                    radius +
+                  center.y;
+      polygon->push_back(toPCL(point.x(), point.y()));
+    }
+  } else if (object.shape.type == Shape::POLYGON) {
+    RCLCPP_WARN_THROTTLE(
+      this->get_logger(), *this->get_clock(), 5, "POLYGON type is not supported");
+  }
+}
+
+std::optional<float> ObstaclePointCloudBasedValidator::getMaxRadius(
+  const autoware_auto_perception_msgs::msg::DetectedObject & object)
+{
+  if (object.shape.type == Shape::BOUNDING_BOX) {
+    return std::hypot(object.shape.dimensions.x * 0.5f, object.shape.dimensions.y * 0.5f);
+  } else if (object.shape.type == Shape::CYLINDER) {
+    return std::hypot(object.shape.dimensions.x * 0.5f, object.shape.dimensions.y * 0.5f);
+  } else if (object.shape.type == Shape::POLYGON) {
+    float max_dist = 0.0;
+    for (const auto & point : object.shape.footprint.points) {
+      const float dist = std::hypot(point.x, point.y);
```

Comment:
I think this is because it follows the same format as visualization_msgs, which supports cylinder, bounding_box, and polygon.
http://docs.ros.org/en/noetic/api/visualization_msgs/html/msg/Marker.html

Reply:
The format is supposed to follow the same format as visualization_msgs CYLINDER, CUBE, TRIANGLE_LIST and etc..
http://docs.ros.org/en/api/visualization_msgs/html/msg/Marker.html
For CYLINDER and CUBE, scale is used and 0.5 is needed like in this case.
For TRIANGLE_LIST is used the relative position from the pose the points.

Reply:
Thanks for the explanation.